src/System.Private.CoreLib/src/System/Threading/WaitSubsystem.HandleManager.Unix.cs

```diff
@@ -0,0 +1,125 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Threading
+{
+    internal static partial class WaitSubsystem
+    {
+        [EagerStaticClassConstruction] // used during lazy class construction
+        private static class HandleManager
+        {
+            private const int InitialCapacity = 4;
+
+            /// <summary>
+            /// Prevents reusing a deleted index until at least there are this many deleted indexes in the free list. The
+            /// purpose is to allow free handles to remain free for some time so that we can discover invalid attempts at using
+            /// free handles.
+            /// </summary>
+            private const int MinimumFreeIndexCountBeforeReuse = 64;
+
+            private static WaitableObject[] s_objects = new WaitableObject[InitialCapacity];
+            private static int s_count;
+            private static LowLevelQueue<int> s_freeIndexes;
```

Comment:
The upper bits of each new handle can be filled with random bits. The `rec.Link != handle` check in FromHandle would detect use of deleted handles because of the upper random bits of the deleted handle won't match upper random bits of the current handle.

Allocation of new handle:
```
int index = s_freeList;
if (index == -1) ...
s_freeList = s_handles[index].Link;

int mask = (s_handles.Length - 1) & 0x3FFFFFFF;
IntPtr handle = s_handles[index] = ((random.Next() & ~mask) | index) + 1;
```